draft-ietf-masque-h3-datagram.md

```diff
@@ -117,26 +123,192 @@ following format (using the notation from the "Notational Conventions" section
 of {{QUIC}}):
 
 ~~~
-HTTP/3 DATAGRAM Frame {
-  Flow ID (i),
+HTTP/3 Datagram {
+  Quarter Stream ID (i),
+  Context ID (i),
   HTTP/3 Datagram Payload (..),
 }
 ~~~
 {: #h3-datagram-format title="HTTP/3 DATAGRAM Frame Format"}
 
-Flow ID:
+Quarter Stream ID:
+
+: A variable-length integer that contains the value of the client-initiated
+bidirectional stream that this datagram is associated with, divided by four.
+(The division by four stems from the fact that HTTP requests are sent on
+client-initiated bidirectional streams, and those have stream IDs that are
+divisible by four.)
+
+Context ID:
+
+: A variable-length integer indicating the context ID of the datagram (see
+{{datagram-contexts}}).
+
+HTTP/3 Datagram Payload:
+
+: The payload of the datagram, whose semantics are defined by individual
+applications. Note that this field can be empty.
+
+Intermediaries parse the Quarter Stream ID field in order to associate the QUIC
+DATAGRAM frame with a stream. If an intermediary receives a QUIC DATAGRAM frame
+whose payload is too short to allow parsing the Quarter Stream ID field, the
+intermediary MUST treat it as an HTTP/3 connection error of type
+H3_GENERAL_PROTOCOL_ERROR. Intermediaries MUST ignore any HTTP/3 Datagram
+fields after the Quarter Stream ID.
```

Comment:
A generic HTTP library capable of receiving and sending datagrams could be used to build either an intermediary or an endpoint.  WIth this specification, It would need two sets of APIs, one that had the context ID (for endpoints) and one without (for intermediaries).  If datagrams carry context IDs then intermediaries should be able to parse them.

Reply:
Your HTTP implementation is going to need to know whether to act on datagrams (endpoint) or forward them (intermediaries). Plumbing that bool to the parse function doesn't seem complex?

Reply:
What would intermediaries do with it? It might be fair to parse it and tell your API users "this field is entirely meaningless unless you are an endpoint".

Reply:
> What would intermediaries do with it?

The same thing it does with other opaque things it get from the peer (eg: body bytes, datagram payloads) - forward to the other peer.

Whatever thing handles an HTTP stream would now have APIs like

```
onDatagram(payload)
sendDatagram(payload)
```
for intermediaries, or, if you are for endpoints:

```
onDatagram(contextID, payload)
sendDatagram(contextID, payload)
```

I'd prefer to have the library only have the second variant.  If using the library to implement a proxy, then the proxy implementation would pass the context ID and payload from one peer's `onDatagram` to the other side's `sendDatagram`. 

Logging/stats are another possible use.

I guess the broader point is -- as this spec is written, the datagram has to have that context ID, so why be so prescriptive about what the intermediary can and can't do with it.

Reply:
We're considering making the Context ID optional (see #45). If intermediaries start parsing the context ID, it'll prevent us from doing that.

Reply:
Whatever mechanism the endpoint will use to know if it should parse the context ID can also be used by the intermediary as well.  My point was that in this PR, they are required.

Presumably if/when negotiation is added the specification can be updated to reflect that.

Reply:
I've filed #54 to keep track of this discussion so it doesn't get lost when this is merged. Let's continue this conversation there.